bgpd/bgp_tx_addpath.c

```diff
@@ -0,0 +1,396 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "bgp_tx_addpath.h"
+#include "bgp_route.h"
+
+static struct txaddpath_strategy_names strat_names[BGP_ADDPATH_MAX] = {
+	{.config_name = "addpath-tx-all-paths",
+	 .human_name = "All",
+	 .human_description = "Advertise all paths via addpath",
+	 .type_json_name = "addpathTxAllPaths",
+	 .id_json_name = "addpathTxIdAll"},
+	{.config_name = "addpath-tx-bestpath-per-AS",
+	 .human_name = "Best-Per-AS",
+	 .human_description = "Advertise bestpath per AS via addpath",
+	 .type_json_name = "addpathTxBestpathPerAS",
+	 .id_json_name = "addpathTxIdBestPerAS"}};
+
+static struct txaddpath_strategy_names unknown_names = {
+	.config_name = "addpath-tx-unknown",
+	.human_name = "Unknown-Addpath-Strategy",
+	.human_description = "Unknown Addpath Strategy",
+	.type_json_name = "addpathTxUnknown",
+	.id_json_name = "addpathTxIdUnknown"};
+
+/*
+ * Returns a structure full of strings associated with an addpath type. Will
+ * never return null.
+ */
+struct txaddpath_strategy_names *
+txaddpath_names(enum bgp_addpath_strategy strat)
+{
+	if (strat < BGP_ADDPATH_MAX) {
+		return &(strat_names[strat]);
+	} else {
+		return &unknown_names;
+	}
+};
+
+/*
+ * Returns if any peer is transmitting addpaths for a given afi/safi.
+ */
+int txaddpath_is_addpath_used(struct txaddpath_bgp_data *d, afi_t afi,
+			      safi_t safi)
+{
+	return d->total_peercount[afi][safi] > 0;
+}
+
+/*
+ * Initialize the BGP instance level data for addpath.
+ */
+void txaddpath_init_bgp_data(struct txaddpath_bgp_data *d)
+{
+	safi_t safi;
+	afi_t afi;
+	int i;
+
+	for (afi = AFI_IP; afi < AFI_MAX; afi++) {
+		for (safi = SAFI_UNICAST; safi < SAFI_MAX; safi++) {
+			for (i = 0; i < BGP_ADDPATH_MAX; i++) {
+				d->id_allocators[afi][safi][i] = NULL;
+				d->peercount[afi][safi][i] = 0;
+			}
+			d->total_peercount[afi][safi] = 0;
+		}
+	}
+}
+
+/*
+ * Free up resources associated with BGP route info structures.
+ */
+void txaddpath_free_info_data(struct txaddpath_info_data *d,
+			      struct txaddpath_node_data *nd)
+{
+	int i;
+	for (i = 0; i < BGP_ADDPATH_MAX; i++) {
+		if (d->addpath_tx_id[i] != IDALLOC_INVALID) {
+			idalloc_free_to_pool(&nd->free_ids[i],
+					     d->addpath_tx_id[i]);
+		}
+	}
+}
+
+/*
+ * Return the addpath ID used to send a particular route, to a particular peer,
+ * in a particular AFI/SAFI.
+ */
+u_int32_t txaddpath_id_for_peer(struct peer *peer, afi_t afi, safi_t safi,
+				struct txaddpath_info_data *d)
+{
+	if (peer->addpath_type[afi][safi] < BGP_ADDPATH_MAX) {
+		return d->addpath_tx_id[peer->addpath_type[afi][safi]];
+	} else {
+		return IDALLOC_INVALID;
+	}
+}
+
+/*
+ * Returns true if the path has an assigned addpath ID for any of the addpath
+ * strategies.
+ */
+int txaddpath_info_has_ids(struct txaddpath_info_data *d)
+{
+	int i;
+	for (i = 0; i < BGP_ADDPATH_MAX; i++) {
+		if (d->addpath_tx_id != 0) {
```

Comment:
bgpd/bgp_tx_addpath.c:122:10: error: comparison of array 'd->addpath_tx_id' not equal to a null pointer is always true [-Werror,-Wtautological-pointer-compare]
                if (d->addpath_tx_id != 0) {
                    ~~~^~~~~~~~~~~~~    ~
1 error generated.